### ISSUE TITLE
LoweringError: Failed at object (object mode backend)

### ISSUE THREAD

**Issue by siyux** [numba/numba#3407]
I want to accelerate the computation of For loops, but I face a bug. Maybe I can't familiar with numba. I just a freshman. But the problem I faced may be useful to improve numba development.
![image](https://user-images.githubusercontent.com/40952612/46912345-83a54500-cfa5-11e8-811a-24ecbc0f3d42.png)
```
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
~/anaconda3/lib/python3.7/site-packages/numba/errors.py in new_error_context(fmt_, *args, **kwargs)
    576     try:
--> 577         yield
    578     except NumbaError as e:

~/anaconda3/lib/python3.7/site-packages/numba/lowering.py in lower_block(self, block)
    253                                    loc=self.loc, errcls_=defaulterrcls):
--> 254                 self.lower_inst(inst)
    255 

~/anaconda3/lib/python3.7/site-packages/numba/objmode.py in lower_inst(self, inst)
     64         if isinstance(inst, ir.Assign):
---> 65             value = self.lower_assign(inst)
     66             self.storevar(value, inst.target.name)

~/anaconda3/lib/python3.7/site-packages/numba/objmode.py in lower_assign(self, inst)
    158         elif isinstance(value, ir.Expr):
--> 159             return self.lower_expr(value)
    160         elif isinstance(value, ir.Global):

~/anaconda3/lib/python3.7/site-packages/numba/objmode.py in lower_expr(self, expr)
    372         else:
--> 373             raise NotImplementedError(expr)
    374 

NotImplementedError: make_function(name=$const0.8, code=<code object <listcomp> at 0x7f1f33a00d20, file "<ipython-input-50-83897177dcb1>", line 10>, closure=None, defaults=None)

During handling of the above exception, another exception occurred:

LoweringError                             Traceback (most recent call last)
<ipython-input-51-8239fdcdd33e> in <module>()
     84                 print("Epoches:{0}:-->".format(j))
     85 
---> 86 mytraining_data,myvalidation_data,mytest_data = load_data_wrapper()
     87 net = NetWork([784,512,28,10])
     88 net.SGD(mytraining_data,10,100,0.1,test_data=mytest_data)

~/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    366                     e.patch_message(''.join(e.args) + help_msg)
    367             # ignore the FULL_TRACEBACKS config, this needs reporting!
--> 368             raise e
    369 
    370     def inspect_llvm(self, signature=None):

~/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    323                 argtypes.append(self.typeof_pyval(a))
    324         try:
--> 325             return self.compile(tuple(argtypes))
    326         except errors.TypingError as e:
    327             # Intercept typing error that may be due to an argument

~/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py in compile(self, sig)
    651 
    652                 self._cache_misses[sig] += 1
--> 653                 cres = self._compiler.compile(args, return_type)
    654                 self.add_overload(cres)
    655                 self._cache.save_overload(sig, cres)

~/anaconda3/lib/python3.7/site-packages/numba/dispatcher.py in compile(self, args, return_type)
     81                                       args=args, return_type=return_type,
     82                                       flags=flags, locals=self.locals,
---> 83                                       pipeline_class=self.pipeline_class)
     84         # Check typing error if object mode is used
     85         if cres.typing_error is not None and not flags.enable_pyobject:

~/anaconda3/lib/python3.7/site-packages/numba/compiler.py in compile_extra(typingctx, targetctx, func, args, return_type, flags, locals, library, pipeline_class)
    871     pipeline = pipeline_class(typingctx, targetctx, library,
    872                               args, return_type, flags, locals)
--> 873     return pipeline.compile_extra(func)
    874 
    875 

~/anaconda3/lib/python3.7/site-packages/numba/compiler.py in compile_extra(self, func)
    365         self.lifted = ()
    366         self.lifted_from = None
--> 367         return self._compile_bytecode()
    368 
    369     def compile_ir(self, func_ir, lifted=(), lifted_from=None):

~/anaconda3/lib/python3.7/site-packages/numba/compiler.py in _compile_bytecode(self)
    802         """
    803         assert self.func_ir is None
--> 804         return self._compile_core()
    805 
    806     def _compile_ir(self):

~/anaconda3/lib/python3.7/site-packages/numba/compiler.py in _compile_core(self)
    789         self.define_pipelines(pm)
    790         pm.finalize()
--> 791         res = pm.run(self.status)
    792         if res is not None:
    793             # Early pipeline completion

~/anaconda3/lib/python3.7/site-packages/numba/compiler.py in run(self, status)
    251                     # No more fallback pipelines?
    252                     if is_final_pipeline:
--> 253                         raise patched_exception
    254                     # Go to next fallback pipeline
    255                     else:

~/anaconda3/lib/python3.7/site-packages/numba/compiler.py in run(self, status)
    243                 try:
    244                     event(stage_name)
--> 245                     stage()
    246                 except _EarlyPipelineCompletion as e:
    247                     return e.result

~/anaconda3/lib/python3.7/site-packages/numba/compiler.py in stage_objectmode_backend(self)
    650         """
    651         lowerfn = self.backend_object_mode
--> 652         self._backend(lowerfn, objectmode=True)
    653 
    654         # Warn if compiled function in object mode and force_pyobject not set

~/anaconda3/lib/python3.7/site-packages/numba/compiler.py in _backend(self, lowerfn, objectmode)
    626             self.library.enable_object_caching()
    627 
--> 628         lowered = lowerfn()
    629         signature = typing.signature(self.return_type, *self.args)
    630         self.cr = compile_result(

~/anaconda3/lib/python3.7/site-packages/numba/compiler.py in backend_object_mode(self)
    599                                      self.library,
    600                                      self.func_ir,
--> 601                                      self.flags)
    602 
    603     def backend_nopython_mode(self):

~/anaconda3/lib/python3.7/site-packages/numba/compiler.py in py_lowering_stage(targetctx, library, interp, flags)
   1016         )
   1017     lower = objmode.PyLower(targetctx, library, fndesc, interp)
-> 1018     lower.lower()
   1019     if not flags.no_cpython_wrapper:
   1020         lower.create_cpython_wrapper()

~/anaconda3/lib/python3.7/site-packages/numba/lowering.py in lower(self)
    171         if self.generator_info is None:
    172             self.genlower = None
--> 173             self.lower_normal_function(self.fndesc)
    174         else:
    175             self.genlower = self.GeneratorLower(self)

~/anaconda3/lib/python3.7/site-packages/numba/lowering.py in lower_normal_function(self, fndesc)
    212         # Init argument values
    213         self.extract_function_arguments()
--> 214         entry_block_tail = self.lower_function_body()
    215 
    216         # Close tail of entry block

~/anaconda3/lib/python3.7/site-packages/numba/lowering.py in lower_function_body(self)
    237             bb = self.blkmap[offset]
    238             self.builder.position_at_end(bb)
--> 239             self.lower_block(block)
    240 
    241         self.post_lower()

~/anaconda3/lib/python3.7/site-packages/numba/lowering.py in lower_block(self, block)
    252             with new_error_context('lowering "{inst}" at {loc}', inst=inst,
    253                                    loc=self.loc, errcls_=defaulterrcls):
--> 254                 self.lower_inst(inst)
    255 
    256     def create_cpython_wrapper(self, release_gil=False):

~/anaconda3/lib/python3.7/contextlib.py in __exit__(self, type, value, traceback)
    128                 value = type()
    129             try:
--> 130                 self.gen.throw(type, value, traceback)
    131             except StopIteration as exc:
    132                 # Suppress StopIteration *unless* it's the same exception that

~/anaconda3/lib/python3.7/site-packages/numba/errors.py in new_error_context(fmt_, *args, **kwargs)
    583         from numba import config
    584         tb = sys.exc_info()[2] if config.FULL_TRACEBACKS else None
--> 585         six.reraise(type(newerr), newerr, tb)
    586 
    587 

~/anaconda3/lib/python3.7/site-packages/numba/six.py in reraise(tp, value, tb)
    657         if value.__traceback__ is not tb:
    658             raise value.with_traceback(tb)
--> 659         raise value
    660 
    661 else:

LoweringError: Failed at object (object mode backend)
make_function(name=$const0.8, code=<code object <listcomp> at 0x7f1f33a00d20, file "<ipython-input-50-83897177dcb1>", line 10>, closure=None, defaults=None)

File "<ipython-input-50-83897177dcb1>", line 10:
def load_data_wrapper():
    <source elided>
    tr_d,va_d,te_d = load_data()
    trainning_inputs = [np.reshape(x,(784,1)) for x in tr_d[0]]
    ^
[1] During: lowering "$0.9 = make_function(name=$const0.8, code=<code object <listcomp> at 0x7f1f33a00d20, file "<ipython-input-50-83897177dcb1>", line 10>, closure=None, defaults=None)" at <ipython-input-50-83897177dcb1> (10)
```

```python
import gzip
import pickle
import random
import numpy as np
from numba import jit
def load_data():
    f = gzip.open('mnist.pkl.gz','rb')
    x,y,z = pickle.load(f,encoding='iso-8859-1')
    print("x = ",len(x))
    print(x)
    return (x,y,z)
@jit
def load_data_wrapper():
    tr_d,va_d,te_d = load_data()
    trainning_inputs = [np.reshape(x,(784,1)) for x in tr_d[0]]
    training_results = [vectorized_result(y) for y in tr_d[1]]
    training_data = [(x,y) for x,y in zip(trainning_inputs,training_results)]
    validation_inputs = [np.reshape(x,(784,1)) for x in va_d[0]]
    validation_data = [(x,y) for x,y in zip(validation_inputs,va_d[1])]
    test_inputs = [np.reshape(x,(784,1)) for x in te_d[0]]
    test_data = [(x,y) for x,y in zip(test_inputs,te_d[1])]
    return (training_data,validation_data,test_data)
def vectorized_result(y):
    e = np.zeros((10,1))
    e[y]=1.0
    return e
class NetWork(object):
    @jit
    def __init__(self,sizes):
        self.n_layer = len(sizes)
        self.sizes = sizes
        self.biases = [np.random.randn(y,1) for y in sizes[1:]]
        self.weights = [np.random.randn(y,x) for  x,y in zip(sizes[:-1],sizes[1:])]

    def sigmoid(self,z):
        return 1.0/(1.0+np.exp(-z))

    def sigmoid_prime(self,z):
        return self.sigmoid(z)*(1-self.sigmoid(z))

    @jit
    def feedforward(self,activation):
        for b,w in zip(self.biases,self.weights):
            activation = self.sigmoid(np.dot(w,activation) + b)
        return activation

    def cost(self,activation ,y):
        activation = self.feedforward(activation)
        z = activation - y
        b = z*z
        b = sum(b)
        return b
    def cost_derivative(self,a,y):
        return (a-y)

    @jit
    def backPropagation(self,x,y):
        delta_nabla_b = [np.zeros(b.shape) for b in self.biases]
        delta_nabla_w = [np.zeros(w.shape) for w in self.weights]
        activation = x
        activations=[x]
        zs = []
        for b,w in zip(self.biases, self.weights):
            z = np.dot(w,activation) + b
            zs.append(z)
            activation= self.sigmoid(z)
            activations.append(activation)
        delta = self.cost_derivative(activations[-1],y)*self.sigmoid_prime(zs[-1])
        delta_nabla_b[-1]= delta
        delta_nabla_w[-1]=  np.dot(delta,activations[-2].transpose())

        for l in range(2,self.n_layer):
            z = zs[-l]
            delta = np.dot(self.weights[-l+1].transpose(),delta) * self.sigmoid_prime(z)
            delta_nabla_b[-l]=delta
            delta_nabla_w[-l] = np.dot(delta,activations[-l-1].transpose())
        return (delta_nabla_b,delta_nabla_w)

    @jit
    def update_mini_batch(self,mini_batch,eta):
        sum_delta_nabla_b = [np.zeros(b.shape) for b in self.biases]
        sum_delta_nabla_w = [np.zeros(w.shape)for w in self.weights]
        for x,y in mini_batch:
            delta_nabla_b,delta_nabla_w = self.backPropagation(x,y)
            sum_delta_nabla_b = [db + sdb for db,sdb in zip(delta_nabla_b,sum_delta_nabla_b)]
            sum_delta_nabla_w = [dw + sdw for dw ,sdw in zip(delta_nabla_w,sum_delta_nabla_w)]
        self.biases = [b -(eta/len(mini_batch))*nb for b ,nb in zip(self.biases,sum_delta_nabla_b)]
        self.weights = [w -(eta/len(mini_batch))*nw for w,nw in zip(self.weights,sum_delta_nabla_w)]
    
    @jit
    def evaluate(self,test_data):
        results = [(np.argmax(self.feedforward(x)),y)  for (x,y) in test_data]
        return sum(int(x==y) for (x,y) in results)
    
    @jit
    def SGD(self,training_data,epchos,mini_batch_size,eta,test_data=None):
        if test_data:
            n_test = len(test_data)
        n_data = len(training_data)

        for j in range(epchos):
            random.shuffle(training_data)
            mini_batchs =[training_data[k:k+mini_batch_size] for k in range(0,n_data,mini_batch_size)]
            for mini_batch in mini_batchs:
                self.update_mini_batch(mini_batch,eta)
            if test_data:
                y_result = self.evaluate(test_data)
                print("Epoches :{0}-->{1}/{2}".format(j,y_result,n_test))
            else:
                print("Epoches:{0}:-->".format(j))

mytraining_data,myvalidation_data,mytest_data = load_data_wrapper()
net = NetWork([784,512,28,10])
net.SGD(mytraining_data,10,100,0.1,test_data=mytest_data)



#net = NetWork([2,3,3,5])
#w = net.pr()


#print(len(w))
#print(w[0])
```
[mnist.pkl.gz](https://github.com/numba/numba/files/2476068/mnist.pkl.gz)

**Comment by stuartarchibald** [comment]
NOTE: edited report into markdown.

**Comment by stuartarchibald** [comment]
Thanks for the report. Please could you confirm which version of Numba you are using, `numba.__version__` will report this. There was a bug fix that went into the v0.40.0 release that I think may fix the problem you are seeing (https://github.com/numba/numba/pull/3122).

**Comment by stuartarchibald** [comment]
Great, please do give it a try and let us know how you get on. It is available via both conda and pip.

**Comment by stuartarchibald** [comment]
Closing, assuming fixed by 0.40.0.
